File: web-core/src/main/java/cn/labzen/web/spring/runtime/LabzenRestRequestHandlerInterceptor.java
package cn.labzen.web.spring.runtime;

import cn.labzen.meta.Labzens;
import cn.labzen.tool.util.Strings;
import cn.labzen.web.api.definition.APIVersionCarrier;
import cn.labzen.web.meta.WebCoreConfiguration;
import cn.labzen.web.util.ControllerDisposeHelper;
import jakarta.annotation.Nonnull;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpStatus;
import org.springframework.web.servlet.HandlerInterceptor;

import java.util.function.Supplier;

/**
 * 请求拦截器
 * <p>
 * 在请求处理之前执行，用于：
 * <ul>
 *   <li>强制验证 API 版本 Header（当配置要求时）</li>
 *   <li>记录请求开始时间，用于统计处理时长</li>
 * </ul>
 */
public class LabzenRestRequestHandlerInterceptor implements HandlerInterceptor {

  /**
   * 判断是否强制要求版本 Header
   * <p>
   * 条件：版本携带方式为 HEADER 且启用了强制验证。
   */
  private final Supplier<Boolean> forceRequestWithVersionHeader = () -> {
    var configuration = Labzens.configurationWith(WebCoreConfiguration.class);
    return configuration.apiVersionCarrier() == APIVersionCarrier.HEADER &&
           configuration.apiVersionHeaderAcceptForced();
  };

  /**
   * 前置处理
   * <p>
   * <ul>
   *   <li>1. 验证 Accept Header 是否包含有效的 API 版本信息</li>
   *   <li>2. 记录请求开始时间</li>
   * </ul>
   */
  @Override
  public boolean preHandle(@Nonnull HttpServletRequest request,
                           @Nonnull HttpServletResponse response,
                           @Nonnull Object handler) {
    // 强制请求的Header中带有Accept版本信息
    if (forceRequestWithVersionHeader.get()) {
      String acceptHeader = request.getHeader("Accept");
      if (acceptHeader == null || !Strings.startsWith(acceptHeader, false, "application/vnd")) {
        response.setStatus(HttpStatus.NOT_ACCEPTABLE.value());
        response.setHeader("Content-Type", "text/plain;charset=UTF-8");
        try {
          response.getWriter().write("Accept header must be specified with a valid API version");
        } catch (Exception e) {
          // 处理异常
        }
        return false;
      }
    }

    ControllerDisposeHelper.recordRequestTime(request);
    return true;
  }
}

File: src/main/kotlin/cn/labzen/web/spring/runtime/LabzenRestRequestHandlerInterceptor.kt
package cn.labzen.web.spring.runtime

import cn.labzen.meta.Labzens
import cn.labzen.tool.util.DateTimes
import cn.labzen.tool.util.Strings
import cn.labzen.web.REST_REQUEST_TIME
import cn.labzen.web.REST_REQUEST_TIME_MILLIS
import cn.labzen.web.defination.APIVersionCarrier
import cn.labzen.web.meta.WebConfiguration
import org.springframework.http.HttpStatus
import org.springframework.web.servlet.HandlerInterceptor
import org.springframework.web.servlet.ModelAndView
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 拦截器，用于统计请求处理时长、以及其他与 Labzen Web 组件相关的功能处理
 */
class LabzenRestRequestHandlerInterceptor : HandlerInterceptor {

  private val forceRequestWithVersionHeader: Boolean by lazy {
    val configuration = Labzens.configurationWith(WebConfiguration::class.java)
    configuration.apiVersionCarrier() == APIVersionCarrier.HEADER && configuration.apiVersionHeaderAcceptForced()
  }

  override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
    // 强制请求的Header中带有Accept版本信息
    if (forceRequestWithVersionHeader) {
      val acceptHeader = request.getHeader("Accept")
      if (acceptHeader == null || !Strings.startsWith(acceptHeader, false, "application/vnd")) {
        response.sendError(
          HttpStatus.NOT_ACCEPTABLE.value(),
//          HttpServletResponse.SC_NOT_ACCEPTABLE,
          "Accept header must be specified with a valid API version"
        )
        return false
      }
    }

    // 用于统计一个请求的用时
    request.setAttribute(REST_REQUEST_TIME_MILLIS, System.currentTimeMillis())
    request.setAttribute(REST_REQUEST_TIME, DateTimes.formatNow())
    return true
  }

  override fun postHandle(
    request: HttpServletRequest,
    response: HttpServletResponse,
    handler: Any,
    modelAndView: ModelAndView?
  ) {
  }

  override fun afterCompletion(
    request: HttpServletRequest,
    response: HttpServletResponse,
    handler: Any,
    ex: Exception?
  ) {
  }

}